### PR TITLE
Add CEPHFS support

### LINKED ISSUE #4822: Ceph FS support
Hello. LXD 3.2 supports Ceph RBD as a storage backend. It will be very good to add a support for pure Ceph (CephFS).
Unfortunately, there is no support of such FS, so I have to mount my CephFS to /mnt/ceph
`sudo mount -t ceph 192.168.0.1:6789:/ /mnt/ceph -o name=admin,secretfile=/etc/ceph/admin.secret`
 and then create an additional dir storage pool
`lxc storage create cephfs dir source=/mnt/ceph`

### PR DESCRIPTION
**Pull request by stgraber**
Closes #4822

Signed-off-by: Stéphane Graber <stgraber@ubuntu.com>

### CODE REVIEW

**Comment by stgraber** [comment]
Ready for review

File: lxd/storage_cephfs.go [hunk 1]
```diff
@@ -0,0 +1,1060 @@
...
+
+func (s *storageCephFs) StorageCoreInit() error {
+	s.sType = storageTypeCeph
+	typeName, err := storageTypeToString(s.sType)
+	if err != nil {
+		return err
+	}
+	s.sTypeName = typeName
+
+	if cephVersion != "" {
```
Comment (brauner) [inline lxd/storage_cephfs.go]: Where does this variable come from?

File: lxd/storage_cephfs.go [hunk 2]
```diff
@@ -0,0 +1,1060 @@
...
+
+	// set osd pool name
+	if s.pool.Config["cephfs.path"] != "" {
+		s.FsName = s.pool.Config["cephfs.path"]
+	}
+
+	return nil
+}
+
+// Initialize a full storage interface.
```
Comment (brauner) [inline lxd/storage_cephfs.go]: That comment seems wrong. :)

File: lxd/storage_cephfs.go [hunk 3]
```diff
@@ -0,0 +1,1060 @@
...
+}
+
+func (s *storageCephFs) StoragePoolCreate() error {
+	logger.Infof(`Creating CEPHFS storage pool "%s" in cluster "%s"`, s.pool.Name, s.ClusterName)
+
+	// Setup config
+	s.pool.Config["volatile.initial_source"] = s.pool.Config["source"]
+
+	if s.pool.Config["source"] == "" {
+		return fmt.Errorf("A ceph fs name OR name/path source is required")
```
Comment (brauner) [inline lxd/storage_cephfs.go]: Hm, for consistency maybe use `cephfs` everywhere?

Comment (brauner) [reply lxd/storage_cephfs.go]: Also, that `OR` looks odd. :)

File: lxd/storage_cephfs.go [hunk 4]
```diff
@@ -0,0 +1,1060 @@
...
+	fields := strings.SplitN(s.FsName, "/", 2)
+	fsName := fields[0]
+	fsPath := "/"
+	if len(fields) > 1 {
+		fsPath = fields[1]
+	}
+
+	// Check that the filesystem exists
+	if !cephFsExists(s.ClusterName, s.UserName, fsName) {
+		return fmt.Errorf("The requested '%v' CEPH fs doesn't exist", fsName)
```
Comment (brauner) [inline lxd/storage_cephfs.go]: `cephfs`, `ceph fs`, or `CEPH fs`? :)

File: lxd/storage_cephfs.go [hunk 5]
```diff
@@ -0,0 +1,1060 @@
...
+	defer removeLockFromMap()
+
+	// Check if already unmounted
+	poolMntPoint := getStoragePoolMountPoint(s.pool.Name)
+	if !shared.IsMountPoint(poolMntPoint) {
+		return false, nil
+	}
+
+	// Unmount
+	err := syscall.Unmount(poolMntPoint, 0)
```
Comment (brauner) [inline lxd/storage_cephfs.go]: Probably should add a `MNT_DETACH` for good measure.

File: lxd/storage_cephfs.go [hunk 6]
```diff
@@ -0,0 +1,1060 @@
...
+func (s *storageCephFs) SetStoragePoolVolumeWritable(writable *api.StorageVolumePut) {
+	s.volume.StorageVolumePut = *writable
+}
+
+func (s *storageCephFs) GetContainerPoolInfo() (int64, string, string) {
+	return s.poolID, s.pool.Name, s.pool.Name
+}
+
+func (s *storageCephFs) StoragePoolUpdate(writable *api.StoragePoolPut, changedConfig []string) error {
+	logger.Infof(`Updating CEPHFS storage pool "%s"`, s.pool.Name)
```
Comment (brauner) [inline lxd/storage_cephfs.go]: `cephfs`, `ceph fs`, `CEPH fs`, or `CEPHFS`? :)

Comment (brauner) [reply lxd/storage_cephfs.go]: The list keeps growing.

File: lxd/storage_cephfs.go [hunk 7]
```diff
@@ -0,0 +1,1060 @@
...
+func (s *storageCephFs) ContainerCreate(container container) error {
+	return fmt.Errorf("CEPHFS cannot be used for containers")
+}
+
+func (s *storageCephFs) ContainerCreateFromImage(container container, imageFingerprint string, tracker *ioprogress.ProgressTracker) error {
+	return fmt.Errorf("CEPHFS cannot be used for containers")
+}
+
+func (s *storageCephFs) ContainerCanRestore(container container, sourceContainer container) error {
+	return fmt.Errorf("CEPHfs cannot be used for containers")
```
Comment (brauner) [inline lxd/storage_cephfs.go]: `cephfs`, `ceph fs`, `CEPH fs`, `CEPHFS`, or `CEPHfs`? :D

File: lxd/storage_cephfs.go (hunk 1 above)
Comment (stgraber) [reply lxd/storage_cephfs.go]: It comes from the CEPH storage backend.

**Pull request merged**